torch/distributed/_shard/checkpoint/resharding.py

```diff
@@ -0,0 +1,198 @@
+from typing import List, Tuple
+
+import torch
+from torch.distributed._shard.sharded_tensor import (
+    ShardedTensor,
+)
+from torch.distributed._shard.sharding_spec import (
+    ShardMetadata,
+)
+from torch.distributed._shard.sharding_spec._internals import (
+    _check_shard_metadata_pair_overlap,
+)
+
+from .metadata import (
+    TensorReadRequest,
+    ShardStorageMetadata,
+    ShardedTensorStorageMetadata,
+    TensorWriteRequest,
+)
+
+# This constant is used as the separator character between tensor name and shard name
+STORAGE_KEY_SEPARATOR = "$"
+
+def _shards_get_overlap_region_wrt_saved_tensor(
+    saved_shard: ShardMetadata, current_shard: ShardMetadata
+) -> List[Tuple[int, int, int, int]]:
+    """
+    Return the overlapping region between saved_shard and current_shard.
+    There returned list has the same number of elements as the tensor's dimension.
+    For each element, we produce a tuple with the following contents:
+        (dimension, `saved_shard` offset, `current_shard` offset, length)
+
+    Offsets are relative to each shard.
+    """
+    narrows = []
+    for dim, (
+        saved_shard_offset,
+        current_shard_offset,
+        saved_shard_size,
+        current_shard_size,
+    ) in enumerate(
+        zip(
+            saved_shard.shard_offsets,
+            current_shard.shard_offsets,
+            saved_shard.shard_sizes,
+            current_shard.shard_sizes,
+        )
+    ):
+        min_range_end = min(
+            saved_shard_offset + saved_shard_size,
+            current_shard_offset + current_shard_size,
+        )
+
+        length = min_range_end - max(current_shard_offset, saved_shard_offset)
+
+        if saved_shard_offset > current_shard_offset:
+            offset_for_saved_tensor = 0
+            offset_for_current_tensor = saved_shard_offset - current_shard_offset
+        else:
+            offset_for_saved_tensor = current_shard_offset - saved_shard_offset
+            offset_for_current_tensor = 0
+
+        narrows.append(
+            (dim, offset_for_saved_tensor, offset_for_current_tensor, length)
+        )
+
+    return narrows
+
+
+def _get_shard_storage_key(
+    tensor_storage_key: str,
+    shard: ShardMetadata
+) -> str:
+    shard_key = "_".join([str(i) for i in shard.shard_offsets])
+    return f"{tensor_storage_key}{STORAGE_KEY_SEPARATOR}{shard_key}"
```

Comment:
While this could generate duplicate FQNs, we ensure there's no conflict by collecting all FQNs, hashing them and use a deterministic way to resolve collisions.

You can see that in _create_storage_key.